libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1347,19 +1397,76 @@ string YulUtilFunctions::updateStorageValueFunction(Type const& _type, std::opti
 			("functionName", functionName)
 			("update",
 				_offset.has_value() ?
-					updateByteSliceFunction(_type.storageBytes(), *_offset) :
-					updateByteSliceFunctionDynamic(_type.storageBytes())
+					updateByteSliceFunction(_toType.storageBytes(), *_offset) :
+					updateByteSliceFunctionDynamic(_toType.storageBytes())
 			)
 			("offset", _offset.has_value() ? "" : "offset, ")
-			("prepare", prepareStoreFunction(_type))
+			("prepare", prepareStoreFunction(_toType))
 			.render();
 		}
 		else
 		{
-			if (_type.category() == Type::Category::Array)
-				solUnimplementedAssert(false, "");
-			else if (_type.category() == Type::Category::Struct)
+			if (_toType.category() == Type::Category::Array)
 				solUnimplementedAssert(false, "");
+			else if (_toType.category() == Type::Category::Struct)
+			{
+				solAssert(_fromType, "");
+				solAssert(_fromType->category() == Type::Category::Struct, "");
+				auto const& fromStructType = dynamic_cast<StructType const&>(*_fromType);
+				auto const& toStructType = dynamic_cast<StructType const&>(_toType);
+				solAssert(fromStructType.structDefinition() == toStructType.structDefinition(), "");
+				solAssert(fromStructType.location() != DataLocation::Storage, "");
+				solUnimplementedAssert(_offset.has_value() && _offset.value() == 0, "");
+
+				Whiskers templ(R"(
+					function <functionName>(slot, value) {
+						<?fromCalldata>
+							let valueMem := <convertToMemory>(value)
```

Comment:
If we process the members individually anyway, then we maybe can also use calldata member access instead of full abi decoding to memory.

Reply:
I agree with that. If I remember well, I think calldata member access is not implemented yet (at least that's what #8085 says). But I can implement calldata member access, shouldn't be too complicated?

Reply:
Sure, then it's probably better to start with calldata member access.

Reply:
Actually it is already implemented by you (#9177)

Reply:
Fixed as suggested